File: prefig/core/arrow.py
import lxml.etree as ET
import logging
import math
import numpy as np
from . import utilities as util
from . import CTM
from . import user_namespace as un

log = logging.getLogger('prefigure')

# Form arrows to be used with a variety of graphical components
# Arrowheads are created as markers and then added to paths

# We'll make arrow heads for tactile diagrams.  Our arrow heads
# will be created in default SVG coordinates since the outlines
# are expressed in those coordinates.  Otherwise, the outlines 
# grow extremely large and blot out elements in the background.
# Because of this, we need to create an arrow head for each stroke
# width.
def add_tactile_arrowhead_marker(diagram, path, mid=False):
    # get the stroke width from the graphical component
    stroke_width_str = path.get('stroke-width', '1')
    stroke_width = int(stroke_width_str)
    id = 'arrow-head-'+stroke_width_str

    # if we've seen this already, there's no need to create it again
    if diagram.has_reusable(id):
        return id

    # Now we'll construct the regular (un-outlined) arrow head.
    # "angle" below is half of the angle at the tip of the head
    # BANA guidelines say this should be between 15 and 22.5 so
    # our angle is a bit too big, but this angle creates a bit
    # more differentiation between the arrow head and the path
    # it's attached to
    angle = 25
    A = math.radians(angle)  # angle in radians
    # We'll construct the arrow head as if the stroke-width is t=1
    # and then scale it below.  We need to think of the path as a
    # rectangle whose height is 2t.  So the tip of the arrow head
    # needs to extend between the endpoint of the path.  
    # The variable l measures the extent the tip extends beyond
    # this endpoint.  The variable s measures the horizontal extent
    # of the arrow head and y is the vertical extent above the
    # center line of the path
    t = 1
    s = 9
    l = t/math.tan(A)+0.1
    y = s*math.tan(A)

    # Scale and translate to fit the stroke-width
    # p1, p2, and p3 are the vertices of the triangle
    # with p1 as the tip
    ctm = CTM.CTM()
    ctm.scale(stroke_width, stroke_width)
    ctm.translate(s-l, y)
    p1 = ctm.transform((l,0))
    p2 = ctm.transform((l-s,y))
    p3 = ctm.transform((l-s,-y))
    cmds = ['M ' + util.pt2str(p1)]
    cmds.append('L ' + util.pt2str(p2))
    cmds.append('L ' + util.pt2str(p3))
    cmds.append('Z')

    d = ' '.join(cmds)

    # Now we're done constructing the shape of the arrow head
    # so we'll add it to a marker and save as a reusable
    x2 = l-s
    dims = [1, 2*y]

    marker = ET.Element('marker')
    marker.set('id', id)
    marker.set('markerWidth', util.float2str(stroke_width*(l-x2)))
    marker.set('markerHeight', util.float2str(stroke_width*dims[1]))
    marker.set('markerUnits', 'userSpaceOnUse')  # userSpaceOnUse?
    marker.set('orient', 'auto-start-reverse')
    marker.set('refX', util.float2str(stroke_width*abs(x2)))
    marker.set('refY', util.float2str(stroke_width*dims[1]/2))

    ET.SubElement(marker, 'path', attrib=
                  {'d': d,
                   'fill': 'context-stroke',
                   'stroke': 'context-none'
                   }
                  )

    diagram.add_reusable(marker)

    outline_width = 9  # 1/8th of an inch
    # We've finished creating the visible arrow head.  Now we need to create
    # an outline which extends 1/8th of an inch beyond the arrow head.
    # First we push the top edge outline_width units away
    push_angle = math.radians(90 - angle)
    w = outline_width*np.array((math.cos(push_angle), math.sin(push_angle)))
    q1 = p1 + w
    q2 = p2 + w

    # now the left side
    v = outline_width * np.array((-1,0))
    q3 = p2 + v
    q4 = p3 + v

    # now the bottom edge
    w = np.array((w[0], -w[1]))
    q5 = p3 + w
    q6 = p1 + w

    # Now that we have the vertices of the outlined arrow head, we will 
    # translate the coordinate system to use as a marker
    ctm = CTM.CTM()
    ctm.translate(outline_width, outline_width)
    q1, q2, q3, q4, q5, q6 = [util.pt2str(ctm.transform(p)) for p in [q1, q2, q3, q4, q5, q6]]

    # Now construct the path, put it in a marker, and add as a reusable
    # The id appends "-outline"
    cmds = ['M', q1]
    cmds += ['L', q2]
    cmds += ['A', str(outline_width), str(outline_width),'0','0','1',q3]
    cmds += ['L', q4]
    cmds += ['A', str(outline_width), str(outline_width),'0','0','1',q5]
    cmds += ['L', q6]
    cmds += ['A', str(outline_width), str(outline_width),'0','0','1',q1]
    cmds += ['Z']

    d = ' '.join(cmds)                
    
    marker = ET.Element('marker')
    marker.set('id', id+'-outline')
    marker.set('markerWidth', util.float2str(stroke_width*(l-x2)+2*outline_width))
    marker.set('markerHeight', util.float2str(stroke_width*dims[1]+2*outline_width))
    marker.set('markerUnits', 'userSpaceOnUse')  # userSpaceOnUse?
    marker.set('orient', 'auto-start-reverse')
    marker.set('refX', util.float2str(abs(stroke_width*x2)+outline_width))
    marker.set('refY', util.float2str(stroke_width*dims[1]/2+outline_width))

    ET.SubElement(marker, 'path', attrib=
                  {'d': d,
                   'fill': 'context-stroke',
                   'stroke': 'context-none'
                   }
                  )
    diagram.add_reusable(marker)
    return id

# Arrowhead marker for a non-tactile figure
# This will follow the same logic as the tactile arrow head created
# above, only the shape of a regular arrow head is slightly different
# than a tactile arrow head.
def add_arrowhead_marker(diagram,
                         path,
                         mid=False,
                         arrow_width=None,
                         arrow_angles=None):
    if arrow_width is not None:
        try:
            arrow_width = un.valid_eval(arrow_width)
        except:
            log.error(f"Error parsing arrow-width={arrow_width}")
            return
    if arrow_angles is not None:
        try:
            arrow_angles = un.valid_eval(arrow_angles)
        except:
            log.error(f"Error parsing arrow-angles={arrow_angles}")
            return
    else:
        arrow_angles = (24, 60)

    if diagram.output_format() == 'tactile':
        return add_tactile_arrowhead_marker(diagram, path)
    
    # get the stroke width from the graphical component
    stroke_width_str = path.get('stroke-width', '1')
    stroke_width = int(stroke_width_str)
    stroke_color = path.get('stroke')

    # Dimensions are a bit different if the arrow head is at an
    # end or in the middle of a path
    id_data = f"_{arrow_width}_{arrow_angles[0]}_{arrow_angles[1]}"
    if not mid:
        id = 'arrow-head-end-'+stroke_width_str+id_data+'-'+stroke_color
        if arrow_width is None:
            arrow_width = 4
        dims = (1, arrow_width)
    else:
        id = 'arrow-head-mid-'+stroke_width_str+id_data+'-'+stroke_color
        if arrow_width is None:
            arrow_width = 13/3
        dims = (1, arrow_width) #11/3)

    # If we've already created this one, we'll just move on
    if diagram.has_reusable(id):
        return id

    # Next we'll construct the path defining the arrow head
    t, s = [d/2 for d in dims]
    A, B = [math.radians(angle) for angle in arrow_angles]
    l = t/math.tan(A) + 0.1
    x2 = l - s/math.tan(A)
    x1 = x2 + (s-t)/math.tan(B)

    ctm = CTM.CTM()
    ctm.scale(stroke_width, stroke_width)
    ctm.translate(-x2, s)
    p1 = ctm.transform((l,0))
    p2 = ctm.transform((x2,s))
    p3 = ctm.transform((x1,t))
    p4 = ctm.transform((x1,-t))
    p5 = ctm.transform((x2, -s))

    d = 'M ' + util.pt2str(p1)
    d += 'L ' + util.pt2str(p2)
    d += 'L ' + util.pt2str(p3)
    d += 'L ' + util.pt2str(p4)
    d += 'L ' + util.pt2str(p5)
    d += 'Z'

    # Now put the arrow head in a marker and add as a reusable
    marker = ET.Element('marker', attrib=
                        {'id': id,
                         'markerWidth': util.float2str(stroke_width*(l-x2)),
                         'markerHeight': util.float2str(stroke_width*2*s),
                         'markerUnits': 'userSpaceOnUse',
                         'orient': 'auto-start-reverse',
                         'refX': util.float2str(stroke_width*abs(x2)),
                         'refY': util.float2str(stroke_width*s)
                         }
                        )

    ET.SubElement(marker, 'path', attrib=
                  {'d': d,
                   'fill': stroke_color, #'context-stroke',
                   'stroke': 'none'
                   #'stroke': 'context-none'
                   }
                  )

    diagram.add_reusable(marker)

    outline_width = 2
    # We've finished creating the visible arrow head.  Now we need to create
    # an outline which extends 1/8th of an inch beyond the arrow head.
    # First we push the top edge outline_width units away
    push_angle = math.pi/2 - A
    w = outline_width*np.array((math.cos(push_angle), math.sin(push_angle)))
    q1 = p1 + w
    q2 = p2 + w

    # now the left side
    v = outline_width * np.array((-1,0))
    q3 = p2 + v
    q4 = p5 + v

    # now the bottom edge
    w = np.array((w[0], -w[1]))
    q5 = p5 + w
    q6 = p1 + w

    # Now that we have the vertices of the outlined arrow head, we will 
    # translate the coordinate system to use as a marker
    ctm = CTM.CTM()
    ctm.translate(outline_width, outline_width)
    q1, q2, q3, q4, q5, q6 = [util.pt2str(ctm.transform(p)) for p in [q1, q2, q3, q4, q5, q6]]

    # Now construct the path, put it in a marker, and add as a reusable
    # The id appends "-outline"
    cmds = ['M', q1]
    cmds += ['L', q2]
    cmds += ['A', str(outline_width), str(outline_width),'0','0','1',q3]
    cmds += ['L', q4]
    cmds += ['A', str(outline_width), str(outline_width),'0','0','1',q5]
    cmds += ['L', q6]
    cmds += ['A', str(outline_width), str(outline_width),'0','0','1',q1]
    cmds += ['Z']

    d = ' '.join(cmds)                
    
    marker = ET.Element('marker')
    marker.set('id', id+'-outline')
    marker.set('markerWidth', util.float2str(stroke_width*(l-x2)+2*outline_width))
    marker.set('markerHeight', util.float2str(stroke_width*2*s+2*outline_width))
    marker.set('markerUnits', 'userSpaceOnUse')  # userSpaceOnUse?
    marker.set('orient', 'auto-start-reverse')
    marker.set('refX', util.float2str(abs(stroke_width*x2)+outline_width))
    marker.set('refY', util.float2str(stroke_width*s+outline_width))

    ET.SubElement(marker, 'path', attrib=
                  {'d': d,
                   'fill': 'white', #'context-stroke',
                   'stroke': 'none'
                   #'stroke': 'context-none'
                   }
                  )

    diagram.add_reusable(marker)

    return id

def add_arrowhead_to_path(diagram,
                          location,
                          path,
                          arrow_width=None,
                          arrow_angles=None):
    id = add_arrowhead_marker(diagram,
                              path,
                              location[-3:] == 'mid',
                              arrow_width,
                              arrow_angles)
    path.set(location, r'url(#{})'.format(id))

File: packages/prefig-rust/prefig-core/src/core/arrow.rs
//! Port of prefig/core/arrow.py: arrowhead markers added to paths.

use crate::core::ctm::CTM;
use crate::core::diagram::Diagram;
use crate::core::repeat::epub_clean;
use crate::core::utilities::{float2str, pt2str};
use crate::value::py_str;
use crate::xml::{self, El};

/// Tactile arrowheads follow BANA-inspired dimensions and get an outline.
fn add_tactile_arrowhead_marker(diagram: &mut Diagram, path: &El) -> Option<String> {
    let stroke_width_str = path.borrow().get_or("stroke-width", "1");
    let stroke_width: f64 = stroke_width_str.parse().ok()?;
    let id = diagram.prepend_id_prefix(&format!("arrow-head-{stroke_width_str}"));

    if diagram.has_reusable(&id) {
        return Some(id);
    }

    let angle = 25f64;
    let a = angle.to_radians();
    let t = 1.0;
    let s = 9.0;
    let l = t / a.tan() + 0.1;
    let y = s * a.tan();

    diagram.arrow_lengths.insert(id.clone(), l);

    let mut ctm = CTM::new();
    ctm.scale(stroke_width, stroke_width);
    ctm.translate(s - l, y);
    let p1 = ctm.transform([l, 0.0]);
    let p2 = ctm.transform([l - s, y]);
    let p3 = ctm.transform([l - s, -y]);
    let d = format!(
        "M {} L {} L {} Z",
        pt2str(p1, " "),
        pt2str(p2, " "),
        pt2str(p3, " ")
    );

    let x2 = l - s;
    let dims = [1.0, 2.0 * y];

    let marker = xml::new_element("marker");
    {
        let mut m = marker.borrow_mut();
        m.set("id", &id);
        m.set("markerWidth", &float2str(stroke_width * (l - x2)));
        m.set("markerHeight", &float2str(stroke_width * dims[1]));
        m.set("markerUnits", "userSpaceOnUse");
        m.set("orient", "auto-start-reverse");
        m.set("refX", &float2str(stroke_width * x2.abs()));
        m.set("refY", &float2str(stroke_width * dims[1] / 2.0));
    }
    let head = xml::sub_element(&marker, "path");
    {
        let mut h = head.borrow_mut();
        h.set("d", &d);
        h.set("fill", "context-stroke");
        h.set("stroke", "context-none");
    }
    diagram.add_reusable(&marker);

    // the outline marker, 1/8" beyond the arrowhead
    let outline_width = 9.0;
    let push_angle = (90.0 - angle).to_radians();
    let w = [
        outline_width * push_angle.cos(),
        outline_width * push_angle.sin(),
    ];
    let q1 = [p1[0] + w[0], p1[1] + w[1]];
    let q2 = [p2[0] + w[0], p2[1] + w[1]];
    let q3 = [p2[0] - outline_width, p2[1]];
    let q4 = [p3[0] - outline_width, p3[1]];
    let q5 = [p3[0] + w[0], p3[1] - w[1]];
    let q6 = [p1[0] + w[0], p1[1] - w[1]];

    let mut ctm = CTM::new();
    ctm.translate(outline_width, outline_width);
    let pts: Vec<String> = [q1, q2, q3, q4, q5, q6]
        .iter()
        .map(|&q| pt2str(ctm.transform(q), " "))
        .collect();
    let ow = py_str(outline_width);
    let d = format!(
        "M {} L {} A {ow} {ow} 0 0 1 {} L {} A {ow} {ow} 0 0 1 {} L {} A {ow} {ow} 0 0 1 {} Z",
        pts[0], pts[1], pts[2], pts[3], pts[4], pts[5], pts[0]
    );

    let marker = xml::new_element("marker");
    {
        let mut m = marker.borrow_mut();
        m.set("id", &format!("{id}-outline"));
        m.set(
            "markerWidth",
            &float2str(stroke_width * (l - x2) + 2.0 * outline_width),
        );
        m.set(
            "markerHeight",
            &float2str(stroke_width * dims[1] + 2.0 * outline_width),
        );
        m.set("markerUnits", "userSpaceOnUse");
        m.set("orient", "auto-start-reverse");
        m.set(
            "refX",
            &float2str((stroke_width * x2).abs() + outline_width),
        );
        m.set(
            "refY",
            &float2str(stroke_width * dims[1] / 2.0 + outline_width),
        );
    }
    let outline = xml::sub_element(&marker, "path");
    {
        let mut o = outline.borrow_mut();
        o.set("d", &d);
        o.set("fill", "context-stroke");
        o.set("stroke", "context-none");
    }
    diagram.add_reusable(&marker);
    Some(id)
}

pub fn add_arrowhead_marker(
    diagram: &mut Diagram,
    path: &El,
    mid: bool,
    arrow_width: Option<&str>,
    arrow_angles: Option<&str>,
) -> Option<String> {
    let arrow_width_value = match arrow_width {
        Some(attr) => match diagram.ctx.valid_eval(attr) {
            Ok(v) => v.as_num().ok(),
            Err(_) => {
                log::error!("Error parsing arrow-width={attr}");
                return None;
            }
        },
        None => None,
    };
    let arrow_angles = match arrow_angles {
        Some(attr) => match diagram
            .ctx
            .valid_eval(attr)
            .ok()
            .and_then(|v| v.as_vec_f64().ok())
        {
            Some(v) if v.len() >= 2 => [v[0], v[1]],
            _ => {
                log::error!("Error parsing arrow-angles={attr}");
                return None;
            }
        },
        None => [24.0, 60.0],
    };

    if diagram.output_format() == "tactile" {
        return add_tactile_arrowhead_marker(diagram, path);
    }

    let stroke_width_str = path.borrow().get_or("stroke-width", "1");
    let stroke_width: f64 = stroke_width_str.parse().ok()?;
    let stroke_color = path.borrow().get_or("stroke", "none");

    // Python f-string embeds None or the evaluated number
    let width_str = match arrow_width_value {
        Some(w) => py_str(w),
        None => "None".to_string(),
    };
    let id_data = format!(
        "_{width_str}_{}_{}",
        py_str(arrow_angles[0]),
        py_str(arrow_angles[1])
    );
    let (id, arrow_width) = if !mid {
        (
            format!("arrow-head-end-{stroke_width_str}{id_data}-{stroke_color}"),
            arrow_width_value.unwrap_or(4.0),
        )
    } else {
        (
            format!("arrow-head-mid-{stroke_width_str}{id_data}-{stroke_color}"),
            arrow_width_value.unwrap_or(13.0 / 3.0),
        )
    };
    let id = epub_clean(&diagram.prepend_id_prefix(&id));

    if diagram.has_reusable(&id) {
        return Some(id);
    }

    let dims = [1.0, arrow_width];
    let (t, s) = (dims[0] / 2.0, dims[1] / 2.0);
    let a = arrow_angles[0].to_radians();
    let b = arrow_angles[1].to_radians();
    let l = t / a.tan() + 0.1;
    let x2 = l - s / a.tan();
    let x1 = x2 + (s - t) / b.tan();

    diagram.arrow_lengths.insert(id.clone(), l);

    let mut ctm = CTM::new();
    ctm.scale(stroke_width, stroke_width);
    ctm.translate(-x2, s);
    let p1 = ctm.transform([l, 0.0]);
    let p2 = ctm.transform([x2, s]);
    let p3 = ctm.transform([x1, t]);
    let p4 = ctm.transform([x1, -t]);
    let p5 = ctm.transform([x2, -s]);

    let d = format!(
        "M {}L {}L {}L {}L {}Z",
        pt2str(p1, " "),
        pt2str(p2, " "),
        pt2str(p3, " "),
        pt2str(p4, " "),
        pt2str(p5, " ")
    );

    let marker = xml::new_element("marker");
    {
        let mut m = marker.borrow_mut();
        m.set("id", &id);
        m.set("markerWidth", &float2str(stroke_width * (l - x2)));
        m.set("markerHeight", &float2str(stroke_width * 2.0 * s));
        m.set("markerUnits", "userSpaceOnUse");
        m.set("orient", "auto-start-reverse");
        m.set("refX", &float2str(stroke_width * x2.abs()));
        m.set("refY", &float2str(stroke_width * s));
    }
    let head = xml::sub_element(&marker, "path");
    {
        let mut h = head.borrow_mut();
        h.set("d", &d);
        h.set("fill", &stroke_color);
        h.set("stroke", "none");
    }
    diagram.add_reusable(&marker);

    // the outline marker
    let outline_width = 2.0;
    let push_angle = std::f64::consts::FRAC_PI_2 - a;
    let w = [
        outline_width * push_angle.cos(),
        outline_width * push_angle.sin(),
    ];
    let q1 = [p1[0] + w[0], p1[1] + w[1]];
    let q2 = [p2[0] + w[0], p2[1] + w[1]];
    let q3 = [p2[0] - outline_width, p2[1]];
    let q4 = [p5[0] - outline_width, p5[1]];
    let q5 = [p5[0] + w[0], p5[1] - w[1]];
    let q6 = [p1[0] + w[0], p1[1] - w[1]];

    let mut ctm = CTM::new();
    ctm.translate(outline_width, outline_width);
    let pts: Vec<String> = [q1, q2, q3, q4, q5, q6]
        .iter()
        .map(|&q| pt2str(ctm.transform(q), " "))
        .collect();
    let ow = py_str(outline_width);
    let d = format!(
        "M {} L {} A {ow} {ow} 0 0 1 {} L {} A {ow} {ow} 0 0 1 {} L {} A {ow} {ow} 0 0 1 {} Z",
        pts[0], pts[1], pts[2], pts[3], pts[4], pts[5], pts[0]
    );

    let marker = xml::new_element("marker");
    {
        let mut m = marker.borrow_mut();
        m.set("id", &format!("{id}-outline"));
        m.set(
            "markerWidth",
            &float2str(stroke_width * (l - x2) + 2.0 * outline_width),
        );
        m.set(
            "markerHeight",
            &float2str(stroke_width * 2.0 * s + 2.0 * outline_width),
        );
        m.set("markerUnits", "userSpaceOnUse");
        m.set("orient", "auto-start-reverse");
        m.set(
            "refX",
            &float2str((stroke_width * x2).abs() + outline_width),
        );
        m.set("refY", &float2str(stroke_width * s + outline_width));
    }
    let outline = xml::sub_element(&marker, "path");
    {
        let mut o = outline.borrow_mut();
        o.set("d", &d);
        o.set("fill", "white");
        o.set("stroke", "none");
    }
    diagram.add_reusable(&marker);

    Some(id)
}

pub fn add_arrowhead_to_path(
    diagram: &mut Diagram,
    location: &str,
    path: &El,
    arrow_width: Option<&str>,
    arrow_angles: Option<&str>,
) -> Option<String> {
    let id = add_arrowhead_marker(
        diagram,
        path,
        location.ends_with("mid"),
        arrow_width,
        arrow_angles,
    )?;
    path.borrow_mut().set(location, &format!("url(#{id})"));
    Some(id)
}
